### PR TITLE
[BUG][STACK-1516/STACK-1526] Reworked HMT in response to STACK-1484/1491

### PR DESCRIPTION
**Pull request by hthompson-a10**
## Description
**STACK-1516**
Severity Level: Critical

Setting `enable_hierarchical_multitenacy=True` doesn't override logic surrounding enforcement of partition per project. This made it impossible to store a full hierarchical structure on a single Thunder which is the deployment requirement when `use_parent_partition` is expected to be used.

**STACK-1526**
Severity Level: High

When an L3V partition is removed from the configuration, it can be put in an inactive state. This leads to `exists()` returning as true, but `active-partition` fails.

Included in this PR since the fix is small and was causing failure in HMT testing env.

## Jira Ticket
[STACK-1516](https://a10networks.atlassian.net/browse/STACK-1516)
[STACK-1526](https://a10networks.atlassian.net/browse/STACK-1526)

## Technical Approach
**STACK-1516**
To handle hierarchical multitenancy partition name collisions and Partition per Project logic, the partition name was assigned at the config validation level instead of the database task level. This allows older partition per project validation code to be reused.

In order to determine whether listed partitions are desired to be set with hierarchical mulitenancy, the option has been moved to the local definition. This has the added effect of clearing ambiguity on the operator side though at the cost of additional setup time.

For now,`use_parent_partition`, a sub-option for hierarchical mulitenancy has been kept at the global level for the time being. It was intended as a one-off switch which is used once and then disabled.

**STACK-1526**
During manual testing, it was found that half-deleted partitions broke HMT (and assumedly l3v partition logic as a whole). To handle half-deleted partitions, extra checks are preformed against a partition whose status is "Not-Active".

## Config Changes
<pre>
[hardware_thunder]

devices=[
 {
 "project_id": "fake_uuid",
 "username" : "username",
 "password" : "password",
 "ip_address" : "10.43.12.137",
 "device_name" : "rack_1"
 <b>"hierarchical_multitenancy":  "&lt;enable/disable&gt;",</b>
 },
 }]
</pre>

Hierarchical Multitenancy has moved from `[a10_global]` to a local definition on the device config entry.

## Test Cases
**STACK-1526**

- Given a partition, when it is not found, then a new one is created
- Given a partition, when an ACOSException or connection failure occurs, then the failure is raised
- Given a partition, when it is found, then no further actions are taken
- Given a partition, when it is found but "Not-Active", then a PartitionNotActiveError is raised

### STACK-1516
Below config was used for testing.

```
[a10_controller_worker]
network_driver = a10_octavia_neutron_driver

[listener]
autosnat = True

[hardware_thunder]
devices = [
                    {
                     "project_id": "6e8ab3d60e6248d09a97b76dda362397",
                     "device_name": "device1",
                     "ip_address": "10.0.0.78",
                     "username": "admin",
                     "password": "a10",
                     "hierarchical_multitenancy": "enable"
                     },
                    {
                     "project_id": "9adffd6e4c554830a4bb48e259caf56f",
                     "device_name": "device1",
                     "ip_address": "10.0.0.78",
                     "username": "admin",
                     "password": "a10",
                     "hierarchical_multitenancy": "enable"
                     },
                    {
                     "project_id": "8c589595eb924356bcfb1416f7b35551",
                     "device_name": "device1",
                     "ip_address": "10.0.0.78",
                     "username": "admin",
                     "password": "a10",
                     "partition_name": "PartitionA",
                     "hierarchical_multitenancy": "disable"
                     }
             ]
```

#### Steps
1. Create the 4 openstack following projects: Parent, Child1, Child2, and lastly ProjectA (no hierarchy)
2. Add entries for Child1, Child2, and ProjectA to the a10-octavia.conf
3. Create load balancers on each of the 3 partitions

##### Expected Results 

```
Total Number of active partitions: 3
Partition Name   Id     L3V/SP     Parent L3V           App Type   Admin Count
------------------------------------------------------------------------------
6e8ab3d60e6248   1       L3V       -                    -            0    
9adffd6e4c5548   2       L3V       -                    -            0    
8c589595eb9243   3       L3V       -                    -            0    
```

```
!
active-partition 6e8ab3d60e6248
!
!
!
slb virtual-server 0a0d7627-df95-433e-a3b5-d0e1f94caf6b 18.64.10.101

!
active-partition 9adffd6e4c5548
!
!
!
slb virtual-server 3d0cb11e-039c-4a19-aa01-2f123d73a0bd 18.64.10.34

!
active-partition 8c589595eb9243
!
!
!
slb virtual-server ecbda83d-1598-4c51-9fed-9c01f755eaf3 18.64.10.27
```

### STACK-1526

```
[a10_controller_worker]
network_driver = a10_octavia_neutron_driver

[listener]
autosnat = True

[hardware_thunder]
devices = [
                    {
                     "project_id": "8c589595eb924356bcfb1416f7b35551",
                     "device_name": "device1",
                     "ip_address": "10.0.0.78",
                     "username": "admin",
                     "password": "a10",
                     "partition_name": "PartitionA"
                     }
             ]
```

#### Steps
1. Create partition "PartitionA" on the Thunder device
2. Execute the command `no partition XXXX id X` to partially remove the config
3. Set the partition name to "PartitionA" in the `a10-octavia.conf`
4. Create a loadbalancer

#### Expected Result
`PartitionNotActiveError` exception is raised

### CODE REVIEW

File: a10_octavia/common/utils.py [hunk 1]
```diff
@@ -88,22 +89,28 @@ def convert_to_hardware_thunder_conf(hardware_list):
...
-                                           ' in [hardware_thunder] section')
+        project_id = hardware_device['project_id']
+        if hardware_dict.get(project_id):
+            raise cfg.ConfigFileValueError('Supplied duplicate project_id {} '
+                                           ' in [hardware_thunder] section'.format(project_id))
         hardware_device['undercloud'] = True
         if hardware_device.get('interface_vlan_map'):
             hardware_device['device_network_map'] = validate_interface_vlan_map(hardware_device)
             del hardware_device['interface_vlan_map']
+        hierarchical_mt = hardware_device.get('hierarchical_multitency')
```
Comment (afrin-chakure-a10) [inline a10_octavia/common/utils.py:100]: spelling of `hierarchical_multitency` --> `hierarchical_multitenancy`

File: a10_octavia/common/utils.py [hunk 2]
```diff
@@ -88,22 +89,28 @@ def convert_to_hardware_thunder_conf(hardware_list):
...
+        if hierarchical_mt == "enable":
+            hardware_device["partition_name"] = project_id[0:14]
+        elif hierarchical_mt != "disable" and hierarchical_mt is not None:
+            raise cfg.ConfigFileValueError('Option `hierarchical_multitency` specified '
```
Comment (afrin-chakure-a10) [inline a10_octavia/common/utils.py:104]: spelling should be `hierarchical_multitenancy`

File: a10_octavia/common/utils.py (hunk 1 above)
Comment (hthompson-a10) [reply a10_octavia/common/utils.py:100]: Yikes thanks for catching this. I had typed it in wrong in my config and was copying/pasting out from that

File: a10_octavia/controller/worker/tasks/vthunder_tasks.py [hunk 3]
```diff
@@ -352,15 +352,22 @@ class HandleACOSPartitionChange(VThunderBaseTask):
...
+        if partition is None:
+            try:
                 axapi_client.system.partition.create(vthunder.partition_name)
                 axapi_client.system.action.write_memory(partition="shared")
                 LOG.info("Partition %s created", vthunder.partition_name)
-        except (acos_errors.ACOSException, req_exceptions.ConnectionError) as e:
-            LOG.exception("Failed to create parition on vThunder: %s", str(e))
-            raise e
+            except (acos_errors.ACOSException, req_exceptions.ConnectionError) as e:
+                LOG.exception("Failed to create parition on vThunder: %s", str(e))
```
Comment (afrin-chakure-a10) [inline a10_octavia/controller/worker/tasks/vthunder_tasks.py:367]: typo `partition` (Though I know its not your change)

**Pull request merged**